### ISSUE TITLE
Allow padding streams

### ISSUE THREAD

**Issue by vasilvv** [moq-wg/moq-transport#534]
Since we currently don't allow unknown frame types and unknown stream types, there is no way to inject padding on the application level.  I believe it would be useful to allow for at least padding unidirectional streams (i.e. streams where receiver ignores everything past the stream type), those could be used for probing (e.g. sending a unidirectional stream at the lowest priority to determine if the connection could sustain more application data).

**Comment by ianswett** [comment]
As a publisher, one can publish a track like what you're describing, but it sounds like you want it built into the transport.

Understanding how it might be used would be helpful.  Is the subscriber trying to see if it has more bandwidth so it can increase quality?  If so, why not just subscribe to the higher quality at a lower priority?

**Comment by suhasHere** [comment]
Agree with Ian . We already do a mix of these things today ( hi/low tracks, low priority probe-like stuff) .. May be I am lacking to understand the requirement here ?

**Comment by vasilvv** [comment]
> Understanding how it might be used would be helpful. Is the subscriber trying to see if it has more bandwidth so it can increase quality? If so, why not just subscribe to the higher quality at a lower priority?

Those are valid if you have one publisher and a lot of subscriber, but for 1:1 scenarios, you would typically have the sender produce only one video stream (and then adjust its bitrate unilaterally), so that's not really an option.

**Comment by ianswett** [comment]
For 1:1 scenarios where the sender is only producing a single stream, the publisher/sender can decide to probe for more bandwidth unilaterally.  The simplest but effective approach is to just retransmit data you've already sent, aka world's simplest FEC.

**Comment by afrind** [comment]
There was a discussion in the slack channel about this issue, which AI has summarized thus:

The discussion revolves around ensuring the congestion controller isn't app-limited in BBR (Bottleneck Bandwidth and Round-trip propagation time) congestion control algorithm. @kixelated mentions sending QUIC PADDING or PING frames periodically, specifically during the BBR ProbeBW_UP phase, to fill the congestion window until a target bitrate is achieved.

For WebTransport clients, creating a "padding" stream with a high sendOrder (lowest priority) is suggested. However, timing the padding with ProbeBW_UP (~1/8th of the time) is crucial for it to be effective with BBR. @vasilvv mentions BBRv1 adjusts cycles depending on whether you're app-limited or not, and ensuring the probe is at least 8 RTTs long can also help.

The goal of padding is to increase bandwidth, and it should be used when the padding rate is smaller than the maximum bitrate. Constant padding can minimize wasted bandwidth, but padding only during ProbeBW_UP is more efficient.

**Comment by kixelated** [comment]
The danger of application padding is that the underlying QUIC library MUST deprioritize any padding stream otherwise it'll start impacting the broadcast. As I result, don't think you should use padding streams today with WebTransport (upload) because `sendOrder` is only a hint and not even implemented in Chrome yet. It's highly dependent on the underlying transport too; I don't think you want padding streams with WebTransport over HTTP/2 or QMux.

Additionally, QUIC PADDING frames are not subject to flow control, so you can't accidentally hit MAX_DATA early and block real data. PADDING frames are not retransmitted, unlike application level padding which will be retransmitted with highest priority after congestion.

I think bandwidth probing is the type of feature that requires explicit support from the underlying QUIC stack unfortunately. The timing and duration of padding also depends on the underlying congestion control algorithm.

**Comment by vasilvv** [comment]
> `sendOrder` is only a hint

That isn't entirely true, if there are two streams with different send orders, one of them will starve another, and that is not an optional behavior of the current W3C spec.

> not even implemented in Chrome yet

That is correct, though MoQ already does rely on send order working in order to operate well under poor network conditions.

> I don't think you want padding streams with WebTransport over HTTP/2 or QMux.

Why not?

> bandwidth probing is the type of feature that requires explicit support from the underlying QUIC stack unfortunately

Well, we don't have this feature guaranteed to be in every QUIC strack, and given we do need some form of probing up, I am not sure that saying "guess we just can't do this" and giving up at this point is an option.

**Comment by ianswett** [comment]
The WG was largely supportive of closing this with no action, though Victor commented that I didn't fully explain what he was trying to describe in this issue, because a Track that's all padding is different and slightly more costly than streams where the payload is discarded.

Unless we're also going to add a way to request the proposed "Padding" streams be sent (ie: a control message?), I'm unsure how we could use them in MoQ?

**Comment by afrind** [comment]
@vasilvv this issue is blocked on you.

**Comment by vasilvv** [comment]
My proposal is:
- Add a new unidirectional stream type.
- The stream has to be all zeroes.
- The receiver has to discard all data on the stream.

That's what we currently do in our implementation, at least.

**Comment by wilaw** [comment]
> My proposal is:
> 
> * Add a new unidirectional stream type.
> * The stream has to be all zeroes.
> * The receiver has to discard all data on the stream.

I support this approach. The advantage of enabling this versus an application-provided stream of zeroes, is that the application provided stream must be routed from the original publisher all the way through the network to the edge, when its really only needed at the dge. This is wasted mid-gress traffic and midgress cache. Being able to have the edge relay spawn this stream would be more efficient. I would extend the design to enable the Subscription to this special stream to specify the desired bitrate i.e SUBSCRIBE (padding, 1.2Mbps)

**Comment by afrind** [comment]
@vasilvv can you write a PR for this?

**Comment by ianswett** [comment]
I left a comment on the PR, but I'm not sure if a stream is the best option due to retransmissions?  Are datagrams better?
